File: composeApp/src/androidMain/kotlin/studio/nxtech/fujubank/features/signup/SignUpComponents.kt
package studio.nxtech.fujubank.features.signup

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import studio.nxtech.fujubank.R

/**
 * サインアップ 3 画面で共通利用する UI 部品群。
 *
 * 既存 LoginScreen でも類似のパーツが組まれているが、A2f では UI のみで API/Koin 依存がないため
 * features/signup 配下にローカルな private 想定で寄せる。LoginScreen 側の移行は別タスク扱い。
 */

internal object SignUpTokens {
    val Background = Color(0xFFF6F7F9)
    val Card = Color.White
    val Primary = Color(0xFFFF1E9E)
    val PrimaryText = Color(0xFF111111)
    val SecondaryText = Color(0xFF6E6F72)
    val Placeholder = Color(0xFFDADBDF)
    val SubText = Color(0xFF64748B)
    val LinkBlue = Color(0xFF006CD7)
    val DividerLine = Color(0xFFE9E9EC)
    val DividerLabel = Color(0xFFC5C5CB)
    val IndicatorActive = Color(0xFF111111)
    val IndicatorInactive = Color(0xFFE1E2E4)
    val OtpUnderlineActive = Color(0xFF333436)
    val OtpUnderlineIdle = Color(0xFFE8E9ED)
    val CtaDisabledBg = Color(0xFFE6E6E6)
    val CtaDisabledText = Color(0xFFC3C3CA)
}

@Composable
internal fun SignUpHeader(
    onBack: (() -> Unit)?,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 8.dp)
            .height(48.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Box(
            modifier = Modifier
                .size(48.dp)
                .let { if (onBack != null) it.clickable(onClick = onBack) else it },
            contentAlignment = Alignment.Center,
        ) {
            if (onBack != null) {
                Icon(
                    painter = painterResource(R.drawable.ic_chevron_left),
                    contentDescription = "戻る",
                    tint = SignUpTokens.PrimaryText,
                    modifier = Modifier.size(24.dp),
                )
            }
        }
        Image(
            painter = painterResource(R.drawable.fuju_wordmark),
            contentDescription = "fuju pay",
            modifier = Modifier.height(28.dp),
            contentScale = ContentScale.Fit,
        )
        Spacer(modifier = Modifier.size(48.dp))
    }
}

@Composable
internal fun BankTextField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    keyboardType: KeyboardType,
    isPassword: Boolean = false,
    enabled: Boolean = true,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(48.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(SignUpTokens.Card),
        contentAlignment = Alignment.CenterStart,
    ) {
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            enabled = enabled,
            singleLine = true,
            cursorBrush = SolidColor(SignUpTokens.PrimaryText),
            visualTransformation = if (isPassword) PasswordVisualTransformation() else VisualTransformation.None,
            keyboardOptions = KeyboardOptions(keyboardType = keyboardType),
            textStyle = LocalTextStyle.current.copy(
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = SignUpTokens.PrimaryText,
            ),
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 24.dp),
            decorationBox = { innerTextField ->
                if (value.isEmpty()) {
                    Text(
                        text = placeholder,
                        style = TextStyle(
                            fontSize = 14.sp,
                            fontWeight = FontWeight.SemiBold,
                            color = SignUpTokens.Placeholder,
                        ),
                    )
                }
                innerTextField()
            },
        )
    }
}

@Composable
internal fun PrimaryButton(
    text: String,
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Button(
        onClick = onClick,
        enabled = enabled,
        shape = RoundedCornerShape(16.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = SignUpTokens.Primary,
            contentColor = Color.White,
            disabledContainerColor = SignUpTokens.CtaDisabledBg,
            disabledContentColor = SignUpTokens.CtaDisabledText,
        ),
        border = BorderStroke(0.dp, Color.Transparent),
        modifier = modifier.height(48.dp),
    ) {
        Text(
            text = text,
            style = TextStyle(
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
            ),
        )
    }
}

@Composable
internal fun DividerWithLabel(label: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Box(
            modifier = Modifier
                .weight(1f)
                .height(2.dp)
                .clip(RoundedCornerShape(27.dp))
                .background(SignUpTokens.DividerLine),
        )
        Text(
            text = label,
            style = TextStyle(
                fontSize = 12.sp,
                fontWeight = FontWeight.Normal,
                color = SignUpTokens.DividerLabel,
            ),
        )
        Box(
            modifier = Modifier
                .weight(1f)
                .height(2.dp)
                .clip(RoundedCornerShape(27.dp))
                .background(SignUpTokens.DividerLine),
        )
    }
}

@Composable
internal fun GoogleSignInButton(onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(48.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(SignUpTokens.Card)
            .clickable(onClick = onClick),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        Image(
            painter = painterResource(R.drawable.google_g),
            contentDescription = null,
            modifier = Modifier.size(20.dp),
        )
        Spacer(Modifier.size(16.dp))
        Text(
            text = "Googleで続ける",
            style = TextStyle(
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
                color = SignUpTokens.PrimaryText,
            ),
        )
    }
}

/**
 * Figma 準拠のページインジケータ。アクティブは 35x6 の角丸ピル、非アクティブは 6x6 の円。
 *
 * Figma の指定通り画面1のみ 4 ドット、画面2/3 は 3 ドットで表現する（実 Pager は使わない）。
 */
@Composable
internal fun PageIndicator(
    total: Int,
    activeIndex: Int,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        repeat(total) { index ->
            if (index == activeIndex) {
                Box(
                    modifier = Modifier
                        .width(35.dp)
                        .height(6.dp)
                        .clip(RoundedCornerShape(20.dp))
                        .background(SignUpTokens.IndicatorActive),
                )
            } else {
                Box(
                    modifier = Modifier
                        .size(6.dp)
                        .clip(RoundedCornerShape(50))
                        .background(SignUpTokens.IndicatorInactive),
                )
            }
        }
    }
}

@Composable
internal fun LoginRedirectLink(onLoginClick: () -> Unit) {
    val annotated = buildAnnotatedString {
        withStyle(
            SpanStyle(
                color = SignUpTokens.SecondaryText,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
            ),
        ) {
            append("アカウントをお持ちの方は ")
        }
        withStyle(
            SpanStyle(
                color = SignUpTokens.Primary,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
                textDecoration = TextDecoration.Underline,
            ),
        ) {
            append("ログイン")
        }
    }
    // 文言全体を 1 つのクリック領域にする。Figma 上もリンク部分のみを精緻にヒット領域化する仕様ではない。
    Text(
        text = annotated,
        modifier = Modifier.clickable(onClick = onLoginClick),
    )
}

@Composable
internal fun LegalAgreementText() {
    // 利用規約 / プライバシーポリシーの遷移先実装は本タスクのスコープ外。視覚のみで配線しない。
    val annotated = buildAnnotatedString {
        withStyle(
            SpanStyle(
                color = SignUpTokens.SecondaryText,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
            ),
        ) {
            append("登録することで、")
        }
        withStyle(
            SpanStyle(
                color = SignUpTokens.LinkBlue,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
                textDecoration = TextDecoration.Underline,
            ),
        ) {
            append("利用規約")
        }
        withStyle(
            SpanStyle(
                color = SignUpTokens.SecondaryText,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
            ),
        ) {
            append(" と ")
        }
        withStyle(
            SpanStyle(
                color = SignUpTokens.LinkBlue,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
                textDecoration = TextDecoration.Underline,
            ),
        ) {
            append("プライバシーポリシー")
        }
        withStyle(
            SpanStyle(
                color = SignUpTokens.SecondaryText,
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
            ),
        ) {
            append(" に同意します")
        }
    }
    Text(text = annotated)
}

File: iosApp/iosApp/Features/SignUp/SignUpComponents.swift
import SwiftUI

/// サインアップ 3 画面で共通利用するデザイントークンと UI 部品。
///
/// LoginView との重複は将来の整理対象。本タスクでは features/SignUp に閉じる方針で寄せる。
enum SignUpTokens {
    static let background = Color(red: 0xF6 / 255, green: 0xF7 / 255, blue: 0xF9 / 255)
    static let card = Color.white
    static let primary = Color(red: 0xFF / 255, green: 0x1E / 255, blue: 0x9E / 255)
    static let primaryText = Color(red: 0x11 / 255, green: 0x11 / 255, blue: 0x11 / 255)
    static let secondaryText = Color(red: 0x6E / 255, green: 0x6F / 255, blue: 0x72 / 255)
    static let placeholder = Color(red: 0xDA / 255, green: 0xDB / 255, blue: 0xDF / 255)
    static let subText = Color(red: 0x64 / 255, green: 0x74 / 255, blue: 0x8B / 255)
    static let linkBlue = Color(red: 0x00 / 255, green: 0x6C / 255, blue: 0xD7 / 255)
    static let dividerLine = Color(red: 0xE9 / 255, green: 0xE9 / 255, blue: 0xEC / 255)
    static let dividerLabel = Color(red: 0xC5 / 255, green: 0xC5 / 255, blue: 0xCB / 255)
    static let indicatorActive = Color(red: 0x11 / 255, green: 0x11 / 255, blue: 0x11 / 255)
    static let indicatorInactive = Color(red: 0xE1 / 255, green: 0xE2 / 255, blue: 0xE4 / 255)
    static let otpUnderlineActive = Color(red: 0x33 / 255, green: 0x34 / 255, blue: 0x36 / 255)
    static let otpUnderlineIdle = Color(red: 0xE8 / 255, green: 0xE9 / 255, blue: 0xED / 255)
    static let ctaDisabledBg = Color(red: 0xE6 / 255, green: 0xE6 / 255, blue: 0xE6 / 255)
    static let ctaDisabledText = Color(red: 0xC3 / 255, green: 0xC3 / 255, blue: 0xCA / 255)
}

struct SignUpHeader: View {
    let onBack: (() -> Void)?

    var body: some View {
        HStack {
            ZStack {
                if let onBack {
                    Button(action: onBack) {
                        Image(systemName: "chevron.left")
                            .font(.system(size: 18, weight: .semibold))
                            .foregroundColor(SignUpTokens.primaryText)
                    }
                }
            }
            .frame(width: 48, height: 48)
            Spacer()
            Image("FujuWordmark")
                .resizable()
                .scaledToFit()
                .frame(height: 28)
            Spacer()
            Color.clear.frame(width: 48, height: 48)
        }
        .padding(.top, 8)
    }
}

struct BankTextField: View {
    @Binding var text: String
    let placeholder: String
    var keyboard: UIKeyboardType = .default
    var secure: Bool = false

    var body: some View {
        ZStack(alignment: .leading) {
            if text.isEmpty {
                Text(placeholder)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(SignUpTokens.placeholder)
            }
            Group {
                if secure {
                    SecureField("", text: $text)
                } else {
                    TextField("", text: $text)
                        .textInputAutocapitalization(.never)
                        .autocorrectionDisabled(true)
                        .keyboardType(keyboard)
                }
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundColor(SignUpTokens.primaryText)
        }
        .padding(.horizontal, 24)
        .frame(height: 48)
        .background(SignUpTokens.card)
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }
}

struct PrimaryButton: View {
    let title: String
    let enabled: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.system(size: 16, weight: .semibold))
                .foregroundColor(enabled ? .white : SignUpTokens.ctaDisabledText)
                .frame(maxWidth: .infinity)
                .frame(height: 48)
                .background(enabled ? SignUpTokens.primary : SignUpTokens.ctaDisabledBg)
                .clipShape(RoundedRectangle(cornerRadius: 16))
        }
        .disabled(!enabled)
    }
}

struct DividerWithLabel: View {
    let label: String

    var body: some View {
        HStack(spacing: 10) {
            line
            Text(label)
                .font(.system(size: 12, weight: .regular))
                .foregroundColor(SignUpTokens.dividerLabel)
            line
        }
        .padding(.horizontal, 24)
        .frame(height: 20)
    }

    private var line: some View {
        RoundedRectangle(cornerRadius: 27)
            .fill(SignUpTokens.dividerLine)
            .frame(height: 2)
            .frame(maxWidth: .infinity)
    }
}

struct GoogleSignInButton: View {
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 16) {
                Image("GoogleG")
                    .resizable()
                    .scaledToFit()
                    .frame(width: 20, height: 20)
                Text("Googleで続ける")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(SignUpTokens.primaryText)
            }
            .frame(maxWidth: .infinity)
            .frame(height: 48)
            .background(SignUpTokens.card)
            .clipShape(RoundedRectangle(cornerRadius: 16))
            .shadow(color: Color.black.opacity(0.02), radius: 3.014, x: 0, y: 3.014)
        }
        .buttonStyle(.plain)
    }
}

/// Figma 準拠のページインジケータ。アクティブは 35x6 の角丸ピル、非アクティブは 6x6 の円。
struct PageIndicator: View {
    let total: Int
    let activeIndex: Int

    var body: some View {
        HStack(spacing: 6) {
            ForEach(0..<total, id: \.self) { index in
                if index == activeIndex {
                    RoundedRectangle(cornerRadius: 20)
                        .fill(SignUpTokens.indicatorActive)
                        .frame(width: 35, height: 6)
                } else {
                    Circle()
                        .fill(SignUpTokens.indicatorInactive)
                        .frame(width: 6, height: 6)
                }
            }
        }
    }
}

struct LoginRedirectLink: View {
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 4) {
                Text("アカウントをお持ちの方は")
                    .font(.system(size: 12, weight: .medium))
                    .foregroundColor(SignUpTokens.secondaryText)
                Text("ログイン")
                    .font(.system(size: 12, weight: .medium))
                    .foregroundColor(SignUpTokens.primary)
                    .underline()
            }
        }
        .buttonStyle(.plain)
    }
}

struct LegalAgreementText: View {
    // 利用規約 / プライバシーポリシーの遷移先実装は本タスクのスコープ外。視覚のみで配線しない。
    var body: some View {
        (
            Text("登録することで、")
                .foregroundColor(SignUpTokens.secondaryText)
            + Text("利用規約")
                .foregroundColor(SignUpTokens.linkBlue)
                .underline()
            + Text(" と ")
                .foregroundColor(SignUpTokens.secondaryText)
            + Text("プライバシーポリシー")
                .foregroundColor(SignUpTokens.linkBlue)
                .underline()
            + Text(" に同意します")
                .foregroundColor(SignUpTokens.secondaryText)
        )
        .font(.system(size: 12, weight: .medium))
        .multilineTextAlignment(.center)
    }
}
